File: medium/3713-longest-balanced-substring-i/solution.java
/**
 * 3713. Longest Balanced Substring I
 * Time: O(n²)
 * Space: O(1)
 */
class Solution {
    public int longestBalancedSubstring(String s) {
        int n = s.length();
        int ans = 0;

        for (int i = 0; i < n; i++) {
            int[] freq = new int[26];
            int distinct = 0;
            int maxFreq = 0;

            for (int j = i; j < n; j++) {
                int idx = s.charAt(j) - 'a';
                if (freq[idx] == 0) {
                    distinct++;
                }
                freq[idx]++;
                maxFreq = Math.max(maxFreq, freq[idx]);

                int len = j - i + 1;
                if (distinct * maxFreq == len) {
                    ans = Math.max(ans, len);
                }
            }
        }

        return ans;
    }
}

File: medium/3713-longest-balanced-substring-i/solution.py
class Solution:
    def longestBalanced(self, s: str) -> int:
        """
        3713. Longest Balanced Substring I
        Time: O(n²)
        Space: O(1)
        """
        n = len(s)
        ans = 0

        for i in range(n):
            freq = [0] * 26
            distinct = 0
            max_freq = 0

            for j in range(i, n):
                idx = ord(s[j]) - ord('a')
                if freq[idx] == 0:
                    distinct += 1
                freq[idx] += 1
                max_freq = max(max_freq, freq[idx])

                length = j - i + 1
                if distinct * max_freq == length:
                    ans = max(ans, length)

        return ans
